README.md
# Simples relógio para stalonetray/openbox

Relógio muito simples feito apenas pra suprir a necessidade do usuário saber a hora e ter um calendário simples à disposição

![stalone_tray_clock](http://alexandrecvieira.droppages.com/images/stalone_tray_clock/stalone_tray_clock.png)

Ao clicar no relógio aparece o calendário

![stalone_tray_clock_calendar](http://alexandrecvieira.droppages.com/images/stalone_tray_clock/stalonetray_clock_calendar2.png)

#### Requerimentos

	sudo apt install stalonetray python-gtk2 python-gobject python-configparser python-xlib

#### Instalação

	git clone https://github.com/alexandrecvieira/stalone_tray_clock.git
	cd stalone_tray_clock
	python setup.py build
    sudo python setup.py install --prefix=/usr
	
	# Configurar para iniciar automaticamente no Openbox
	echo "stalone_tray_clock &" >> $HOME/.config/openbox/autostart
	
#### O arquivo de configuração: ~/.stalone_tray_clockrc é gerado automaticamente com as seguintes configurações padrão:

	icon_size:32
	bgcolor:#111
	fontcolor:white
	fontsize:9.5
	font:Ubuntu Regular
	fontweight:bold
	lang:pt
	
##### lang: 

 * **pt**: formata o relógio para português: Sex, 16 de Mar 16:30
 * **en**: formata o relógio para o formato estrangeiro: Fri Mar 16, 04:30 PM 

##### icon_size deve ser o mesmo tamanho de ícone do stalonetray


ChangeLog

2019-07-21  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* stalone_tray_clock (draw_calendar_window): focus-out-event stalone_tray_clock v0.10

2019-05-06  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Formatação de string de config

2019-05-03  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Melhorado o código de formatação das strings

2019-05-02  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Tooltip
	* Iniciando tentativa de criar um tooltip

2019-04-20  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Destruir a janela do calendário no segundo evento, melhorando a usabilidade do calendário

2019-04-15  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Mudado de gobject para thread  e resolvido bug na contagem dos segundos para as atualizações do dia e mes

2019-04-05  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* stalone_tray_clock (set_update_date_interval): interval update
	* Atualização do dia, dia da semana e mês na virada do dia(00:00)

2019-03-11  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* stalone_tray_clock (draw_calendar_window): janela do calendário centralizada em relação ao ponteiro do mouse
	* stalone_tray_clock (draw_calendar_window): close calendar on visibility events(partial and fully obscured)

2019-03-10  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* tirada decoração da janela do calendário e ocultando a janela ao clicar novamente no relógio

2019-03-05  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* setup.py

2019-03-01  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* nomes dos dias da semana no calendário
	* ícone do calendário
	* Install script update
	* formato estrangeiro
	* nomes dos dias e meses de acordo com locale, adicionado calendário

2019-02-28  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Merge remote-tracking branch 'origin/master'
	* solucionado bug ao carregar arquivo de configuração

2019-02-13  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Fix month error

2019-02-12  Alexandre C Vieira <acamargo.vieira@gmail.com>

	* Update
	* Fix install script
	* Install script
	* Remove unused
	* Screenshot
	* Start...


install.sh
#!/bin/bash
basedir=$(pwd)

ls /opt/stalone_tray_clock

if [ $? -ne 0 ] ; then
    sudo mkdir /opt/stalone_tray_clock
fi

sudo cp stalone_tray_clock.py /opt/stalone_tray_clock/

cat $HOME/.config/openbox/autostart > ~/Downloads/temp.tmp
temp=$(sed -n '/stalone_tray_clock &/p' ~/Downloads/temp.tmp)
if [ ${#temp} -gt 0 ] ; then
    echo "--------------------------------------"
    echo "stalone_tray_clock já estava configurado."
    echo "--------------------------------------"
    rm ~/Downloads/temp.tmp
else
    echo "stalone_tray_clock &" >> ~/Downloads/temp.tmp
    cp ~/Downloads/temp.tmp $HOME/.config/openbox/autostart
    rm ~/Downloads/temp.tmp
    echo "--------------------------------------"
    echo "stalone_tray_clock foi configurado."
    echo "--------------------------------------"
fi

ls /usr/bin/stalone_tray_clock

if [ $? -ne 0 ] ; then
    cd /usr/bin/
    sudo ln -s /opt/stalone_tray_clock/stalone_tray_clock.py stalone_tray_clock
fi


stalone_tray_clock.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

__appname__ = "Stalone Tray Clock"
__version__ = "0.1"
__author__  = "Alexandre C Vieira"
__license__ = "GNU GPL 2.0 or later"

import pygtk
pygtk.require('2.0')
import gtk, gobject, datetime, calendar, configparser, io, os, locale
from Xlib import display

UPDATE_INTERVAL = 1  # in Seconds
SHOW_CALENDAR = False

def mousepos():
    """mousepos() --> (x, y) get the mouse coordinates on the screen (linux, Xlib)."""
    data = display.Display().screen().root.query_pointer()._data
    return data["root_x"], data["root_y"]

def day_of_week(date):
    return calendar.day_abbr[date.weekday()].capitalize()

def current_month(monthNumber):
    return calendar.month_abbr[monthNumber].capitalize() 

def draw_window(label):
    widget = gtk.Layout()
    if label != None:
        widget.put(label,0,ICON_SIZE / 4)
    widget.show()
    window = gtk.OffscreenWindow()
    window.set_default_size(ICON_SIZE, ICON_SIZE)
    window.set_opacity(0)
    window.add(widget)
    window.show_all()
    col = gtk.gdk.Color(BGCOLOR)
    if label == None:
        window.present()
    widget.modify_bg(gtk.STATE_NORMAL, col)
    return window

def draw_calendar_window():
    global SHOW_CALENDAR
    SHOW_CALENDAR = True
    vbox = gtk.VBox(False, 5)
    cal = gtk.Calendar()
    cal.set_display_options(gtk.CALENDAR_SHOW_HEADING | gtk.CALENDAR_SHOW_DAY_NAMES)
    halign1 = gtk.Alignment(0.5, 0.5, 0, 0)
    halign1.add(cal)
    valign = gtk.Alignment(0, 1, 0, 0)
    vbox.pack_start(halign1)
    window = gtk.Window()
    if LANG == 'pt':
        window.set_title("Calendário")
    else:
        window.set_title("Calendar")
    window.set_size_request(250, 200)
    window.set_icon(gtk.icon_theme_get_default().load_icon("calendar", 48, 0))
    mouse = mousepos()
    window.move(mouse[0] - 100, mouse[1] + 30)
    window.add(vbox)
    window.connect("destroy", set_show_calendar)
    window.show_all()

def set_show_calendar(gtkobject, data=None):
    global SHOW_CALENDAR
    SHOW_CALENDAR = False
    
def save_config():
    with open(os.path.expanduser('~/.stalone_tray_clockrc'), 'w') as f:
        f.write('%s\n%s\n%s\n%s\n%s\n%s' % ('icon_size' + ':'
                                                + str(ICON_SIZE), 'bgcolor' + ':' + BGCOLOR, 'fontcolor' + ':'
                                                + FONTCOLOR, 'fontsize' + ':' + FONTSIZE, 'font' + ':'
                                                + FONT, 'fontweight' + ':' + FONTWEIGHT))

def cb_allocate(label, allocation):
    label.set_size_request( ICON_SIZE, -1 )

def on_left_click(event):
    global SHOW_CALENDAR
    if not SHOW_CALENDAR:
        draw_calendar_window()
    
class ClockDayOfWeek:
    def __init__(self):
        self.icon = gtk.StatusIcon()
        self.icon.connect( "activate", on_left_click )
        self.dayofweek = day_of_week(datetime.date.today())
        self.label = gtk.Label()
        self.label.set_justify(gtk.JUSTIFY_CENTER)
        if LANG == 'pt':
            self.label.set_markup('<span color=\'' + FONTCOLOR + '\' font_weight=\'' + FONTWEIGHT
                                  + '\' font=\'' + FONT + '\'>' + self.dayofweek + ', </span>')
        else:
            self.label.set_markup('<span color=\'' + FONTCOLOR + '\' font_weight=\'' + FONTWEIGHT
                                  + '\' font=\'' + FONT + '\'>' + self.dayofweek + '</span>')
        self.label.connect( "size-allocate", cb_allocate )
        self.window = draw_window(self.label)
        self.window.connect("damage-event", self.draw_complete_event)
        # self.window.present()

    def draw_complete_event(self, window, event):
        self.icon.set_from_pixbuf(window.get_pixbuf())

class ClockDay:
    def __init__(self):
        self.icon = gtk.StatusIcon()
        self.icon.connect( "activate", on_left_click )
        self.now = datetime.datetime.now()
        self.label = gtk.Label()
        self.label.set_justify(gtk.JUSTIFY_CENTER)
        if LANG == 'pt':
            self.label.set_markup('<span color=\'' + FONTCOLOR + '\' font_weight=\'' + FONTWEIGHT
                                  + '\' font=\'' + FONT + '\'>' + str(self.now.day) + ' de</span>')
        else:
            self.label.set_markup('<span color=\'' + FONTCOLOR + '\' font_weight=\'' + FONTWEIGHT
                                  + '\' font=\'' + FONT + '\'>' + str(self.now.day) + ', </span>')         
        self.label.connect( "size-allocate", cb_allocate )
        self.window = draw_window(self.label)
        self.window.connect("damage-event", self.draw_complete_event)
        # self.window.present()

    def draw_complete_event(self, window, event):
        self.icon.set_from_pixbuf(window.get_pixbuf())

class ClockMonth:
    def __init__(self):
        self.icon = gtk.StatusIcon()
        self.icon.connect( "activate", on_left_click )
        self.month = current_month(datetime.date.today().month)
        self.label = gtk.Label()
        self.label.set_justify(gtk.JUSTIFY_CENTER)
        self.label.set_markup('<span color=\'' + FONTCOLOR + '\' font_weight=\'' + FONTWEIGHT
                              + '\' font=\'' + FONT + '\'>' + self.month + '</span>')
        self.label.connect( "size-allocate", cb_allocate )
        self.window = draw_window(self.label)
        self.window.connect("damage-event", self.draw_complete_event)
        self.window.present()

    def draw_complete_event(self, window, event):
        self.icon.set_from_pixbuf(window.get_pixbuf())

class ClockTime:
    def __init__(self):
        self.icon = gtk.StatusIcon()
        self.icon.connect( "activate", on_left_click )
        self.label = gtk.Label()
        self.label.set_markup('<span color="white">time</span>')
        self.label.set_justify(gtk.JUSTIFY_CENTER)
        self.window = draw_window(self.label)
        self.window.connect("damage-event", self.draw_complete_event)
        self.update_clock()
        gobject.timeout_add(1000 * UPDATE_INTERVAL, self.update_clock)
        self.now = None
                   
    def update_clock(self):
        self.now = datetime.datetime.now()
        if LANG == 'pt':
            self.label.set_markup('<span color=\'' + FONTCOLOR + '\' font_weight=\'' + FONTWEIGHT
                                  + '\' font=\'' + FONT + '\'>' + self.now.strftime("%H:%M") + '</span>')
        else:
            self.label.set_markup('<span color=\'' + FONTCOLOR + '\' font_weight=\'' + FONTWEIGHT
                                  + '\' font=\'' + FONT + '\'>' + self.now.strftime("%I:%M") + '</span>')
        self.label.connect( "size-allocate", cb_allocate )
        self.window.present()
        return True

    def draw_complete_event(self, window, event):
        self.icon.set_from_pixbuf(window.get_pixbuf())

class ClockTimeAMPM:
    def __init__(self):
        self.icon = gtk.StatusIcon()
        self.icon.connect( "activate", on_left_click )
        self.label = gtk.Label()
        self.label.set_markup('<span color="white">time</span>')
        self.label.set_justify(gtk.JUSTIFY_CENTER)
        self.window = draw_window(self.label)
        self.window.connect("damage-event", self.draw_complete_event)
        self.update_clock()
        gobject.timeout_add(1000 * UPDATE_INTERVAL, self.update_clock)
        self.now = None
                   
    def update_clock(self):
        self.now = datetime.datetime.now()
        self.label.set_markup('<span color=\'' + FONTCOLOR + '\' font_weight=\'' + FONTWEIGHT
                              + '\' font=\'' + FONT + '\'>' + self.now.strftime("%p") + '</span>')
        self.label.connect( "size-allocate", cb_allocate )
        self.window.present()
        return True

    def draw_complete_event(self, window, event):
        self.icon.set_from_pixbuf(window.get_pixbuf())

class ClockPadding:
    def __init__(self):
        self.icon = gtk.StatusIcon()
        window = draw_window(None)
        window.connect("damage-event", self.draw_complete_event)

    def draw_complete_event(self, window, event):
        self.icon.set_from_pixbuf(window.get_pixbuf())
       
if __name__ == "__main__":
    try:
        user_config_dir = open(os.path.expanduser('~/.stalone_tray_clockrc'), 'r').read().split('\n')
        ICON_SIZE = int(user_config_dir[0].split(':')[1])
        BGCOLOR = user_config_dir[1].split(':')[1]
        FONTCOLOR = user_config_dir[2].split(':')[1]
        FONTSIZE = user_config_dir[3].split(':')[1]
        FONT = user_config_dir[4].split(':')[1]
        FONTWEIGHT = user_config_dir[5].split(':')[1] # One of 'ultralight', 'light', 'normal', 'bold',
                                                      # 'ultrabold', 'heavy', or a numeric weight
        LANG = user_config_dir[6].split(':')[1]
    except IOError:
        ICON_SIZE = 32
        BGCOLOR = '#111'
        FONTCOLOR = 'white'
        FONTSIZE = '9.5'
        FONT ='Ubuntu Regular'
        FONTWEIGHT = 'bold' # One of 'ultralight', 'light', 'normal', 'bold',
                            # 'ultrabold', 'heavy', or a numeric weight
        LANG = 'pt' # pt or en
        save_config()

    FONT = FONT + ' ' + FONTSIZE
    
    paddingleft = ClockPadding()
    
    if LANG == 'pt':
        appdayofweek = ClockDayOfWeek()
        appday = ClockDay()
        appmonth = ClockMonth()
    else:
        appdayofweek = ClockDayOfWeek()
        appmonth = ClockMonth()
        appday = ClockDay()

    apptime = ClockTime()

    if LANG != 'pt':
        apptimeampm = ClockTimeAMPM()
        
    paddingright = ClockPadding()
    
    gtk.main()
